### ISSUE TITLE
Interop problem: UMD module import default from System.register module

### ISSUE THREAD

**Issue by shrinktofit** [systemjs/systemjs#2064]
Say I have two modules: the caller is an **ESM** format and the callee is an CJS format. I hope the caller module can import and use callee. This is useful because we encourage developers use ESM module, but we don't want them losing the ability to use hundreds of NPM packages.

These scripts's execution environment is browser which may not support ES6, so we can not convert the ESM modules to CJS one. We also don't want to do bundle cause of time spending.

So I have an idea, which:
- Transform callee **CJS** modules into **ESM** and then into **System.register**(We limits that developers should guarateen that these **CJS** modules contains only explicit requires and module.exports);
- Transform caller **ESM** modules into **UMD**(Yes, UMD instead of System.register, we have reasons).

But I found that, in practice, UMD modules cannot import System.register modules, using SystemJS loader. As following example states.

------------

Say I have a caller,  it import default from callee(in commonJS):
```ts
import $protobuf from "protobuf/minimal"; // protobuf/minimal is a NPM modules, in commonJS
```
I transform it to UMD,
```js
// Omitting the header pattern...
function (_exports, _minimal) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports["default"] = _exports.main = void 0;
  _minimal = _interopRequireDefault(_minimal); // BROKEN

  function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
}
```

and the callee's source:
```js
module.exports = require('anything');
```
I transform it to ES6:
```js
import $ from "anything";
const module = { exports: {} }
const exports = module.exports;
exports = $;
export default module.exports; // Key
```
and to System.register:
```js
System.register(["anything"], function (_export, _context) {
  "use strict";

  var _srcIndexMinimal, module, exports;

  return {
    setters: [function (_srcIndexMinimal2) {
      _srcIndexMinimal = _srcIndexMinimal2.default;
    }],
    execute: function () {
      module = {
        exports: {}
      };
      exports = module.exports;
      module.exports = _srcIndexMinimal;

      _export("default", module.exports);
    }
  };
});
```

When executing caller,
```js
 _minimal = _interopRequireDefault(_minimal);
```
cause `_minimal` to be `{ default: { default: /* exports of anything */ } }`.

----------------

We use `@babel/plugin-transform-modules-systemjs` to transform ES6 to System.register.
Whether we should  append `_export("__esModule", { value: true })` to System.register codes to prevent `_interopRequireDefault` from doing wrap for us?

**Comment by joeldenning** [comment]
Try the [`use-default` extra](https://github.com/systemjs/systemjs/blob/master/dist/extras/use-default.js). It was created specifically to faciliate UMD modules importing System.register modules.

The reason for its existence was discussed in https://github.com/systemjs/systemjs/issues/1967. Also note that recently the extra has been marked for deprecation as discussed in https://github.com/systemjs/systemjs/issues/2045. The reason for deprecation was not that it doesn't make this specific interop situation easier, but rather that people thought it was supposed to do other things besides what it's already doing.

In any case, the use-default extra is the thing that will help you.

**Comment by shrinktofit** [comment]
I had been tried with the use-default extra at the beginning.

**Comment by shrinktofit** [comment]
May be is complex for my showcase. I simply the showcase -- the interop problem is that UMD modules cannot import default from System.register module.
- The UMD can be generated using babel;
- The System.register is generated using rollup.

**Comment by joeldenning** [comment]
Can you please provide a minimal github repo or plunkr/codesandbox/codepen that shows the problem? That will make it easier for me to look into this.

**Comment by shrinktofit** [comment]
https://github.com/shrinktofit/systemjs-issue-2064-demo

Hi I have upload the repo. This repos contains the `/caller.js` in UMD module, `/callee.js` in System.register module. Their source can be found at `/source/`.

Please run `npm install` to setup.

Run `npm run start` to server the repo, and access `http://127.0.0.1:2064/index.html` in browser and open DevTools to run the demo.

If you change sources(files in `/source`), run `npm run build` to build them.

**Comment by joeldenning** [comment]
I just ran the code locally. This is actually expected behavior, unfortunately. See https://github.com/systemjs/systemjs/issues/1933 for discussion about this exact situation.

The explanation: the entire javascript ecosystem, including webpack and babel, embraced a convention where if external modules have an `__esModule` flag on them, that they know how to handle default exports for those modules. The `__esModule` property itself is something that was originally proposed by @guybedford. Since that happened, however, it has been decided that NodeJS will not support the `__esModule` flag. In an effort for systemjs@>=2 to be compatible with specifications going forward, it has been decided that SystemJS will no longer automatically apply the `__esModule` flag to modules. The tradeoff is a less convenient compatibility with babel and webpack, but a more standards-compliant systemjs going forward.

In other words - this is not a bug in systemjs, but there is a workaround to make things more convenient (just add `__esModule` to your systemjs modules). I have created https://github.com/shrinktofit/systemjs-issue-2064-demo/pull/1 which demonstrates what the workaround looks like.

I'm closing this issue since I think we've gotten to the bottom of this, but feel free to reopen / comment further if there's more to discuss.
